### ISSUE TITLE
MySQL Database Issue - Connection Refused 

### ISSUE THREAD

**Issue by IoscaA** [supermavster/docker-laravel-8#3]
Hello @supermavster, I tried to run the migrations from my project to the mysql database, but when I run _docker-compose run --rm artisan migrate --seed_ command the system returns with:
_SQLSTATE[HY000] [2002] Connection refused (SQL: select * from information_schema.tables where table_schema = database and table_name = migrations and table_type = 'BASE TABLE')_ .

**.env** 
{
--MYSQL Password--
DB_ROOT_PASSWORD=secret
--Connection DB--
DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=database
DB_USERNAME=username
DB_PASSWORD=secret
}

**source folder .env** 
{                                   
APP_NAME=Laravel
APP_ENV=local
APP_KEY=base64:43WPYT1zCqO7WmZifstlRbNwf0kUErINrXnLZZAELjo=
APP_DEBUG=true
APP_URL=http://localhost

LOG_CHANNEL=stack

DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=database
DB_USERNAME=username
DB_PASSWORD=secret

BROADCAST_DRIVER=log
CACHE_DRIVER=file
QUEUE_CONNECTION=sync
SESSION_DRIVER=database
SESSION_LIFETIME=120

REDIS_HOST=127.0.0.1
REDIS_PASSWORD=null
REDIS_PORT=6379

MAIL_MAILER=smtp
MAIL_HOST=smtp.mailtrap.io
MAIL_PORT=2525
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS=null
MAIL_FROM_NAME="${APP_NAME}"

AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_DEFAULT_REGION=us-east-1
AWS_BUCKET=

PUSHER_APP_ID=
PUSHER_APP_KEY=
PUSHER_APP_SECRET=
PUSHER_APP_CLUSTER=mt1

MIX_PUSHER_APP_KEY="${PUSHER_APP_KEY}"
MIX_PUSHER_APP_CLUSTER="${PUSHER_APP_CLUSTER}"
}

**Comment by supermavster** [comment]
Hello, Remember:

When Docker generate a database, the `DB_HOST` change to the mane of 'container' that you create, in this case is `mysql`:

```
  mysql:
    image: mysql:5.7.29
    .....
```

So... in your **env of the folder source**, you need put: 

```
DB_HOST=mysql
```

Try and tell me if this works for you ;) @IoscaA

**Comment by IoscaA** [comment]
I make that change, but nothing has changed. Also I can't even access the mysql database with phpmyadmin. I don't know what I'm doing wrong, but there's no link between mysql and phpmyadmin.

Also with     _docker-compose run --rm artisan migrate --seed_  I've now:

_SQLSTATE[HY000] [1130] Host 'laravelproject_artisan_run_6148f2fd1a36.laravelproject' is not allowed to connect to this MySQL server (SQL: select * from information_schema.tables where table_schema = database and table_name = migrations and table_type = 'BASE TABLE')_

**Thanks for all you are doing :)**

**Comment by supermavster** [comment]
;) Don't Worry.


Okey, in the files, check the config:

**.env**
```
DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=database
DB_USERNAME=userdb
DB_PASSWORD=passworddb
DB_ROOT_PASSWORD=secret
```

and **Source/.env**
```
DB_CONNECTION=mysql
DB_HOST=mysql
DB_PORT=3306
DB_DATABASE=database
DB_USERNAME=userdb
DB_PASSWORD=passworddb
```

Save the files.

Put the next commands in the shell or terminal to clean the cache and the config data:

```
docker-compose run --rm artisan cache:clear
docker-compose run --rm artisan config:cache
docker-compose run --rm artisan key:generate
```

Try again whit the command:

```
docker-compose run --rm artisan migrate --seed
```

Try and tell me if this works for you ;) @IoscaA

**Comment by IoscaA** [comment]
I replaced both files with your solution, but nothing has changed :'c

Error:  _SQLSTATE[HY000] [1130] Host 'docker-laravel-8_artisan_run_da68afbc55b5.docker-laravel-8_larav' is not allowed to connect to this MySQL server (SQL: select * from information_schema.tables where table_schema = database and table_name = migrations and table_type = 'BASE TABLE')_

**Comment by IoscaA** [comment]
I tried to change the mysql image in the docker-compose.yml to -->  :latest
But I obtain this error if I run: _docker-compose run --rm artisan migrate --seed_

**ERROR**: _SQLSTATE[HY000] [2002] php_network_getaddresses: getaddrinfo failed: Try again (SQL: select * from information_schema.tables where table_schema = database and table_name = migrations and table_type = 'BASE TABLE')_

**Comment by supermavster** [comment]
The error say: `table_name = migrations`

Your database is _migrations_?

--- 

Delete and Build again:

```bash
docker-compose down -v
docker-compose up --build
```

And in a new Shell (Without close the other shell), execute: 
```
docker-compose run --rm artisan migrate --seed
```
Try and tellme ;) @IoscaA

**Comment by IoscaA** [comment]
Now I've this: *S_QLSTATE[HY000] [2002] php_network_getaddresses: getaddrinfo failed: Try again (SQL: select * from information_schema.tables where table_schema = database and table_name = migrations and table_type = 'BASE TABLE')*

On the other terminal:

mysql         | 2020-11-09 17:17:35+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mysql         | 2020-11-09 17:17:35+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.22-1debian10 started.
mysql         | 2020-11-09 17:17:35+00:00 [Note] [Entrypoint]: Initializing database files
mysql         | 2020-11-09T17:17:35.219000Z 0 [Warning] [MY-010915] [Server] 'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode. They will be merged with strict mode in a future release.
mysql         | 2020-11-09T17:17:35.219074Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.22) initializing of server in progress as process 43
mysql         | 2020-11-09T17:17:35.220057Z 0 [Warning] [MY-013242] [Server] --character-set-server: 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous.
mysql         | 2020-11-09T17:17:35.220756Z 0 [ERROR] [MY-010457] [Server] --initialize specified but the data directory has files in it. Aborting.
mysql         | 2020-11-09T17:17:35.220763Z 0 [ERROR] [MY-013236] [Server] The designated data directory /var/lib/mysql/ is unusable. You can remove all files that the server added to it.
mysql         | 2020-11-09T17:17:35.223655Z 0 [ERROR] [MY-010119] [Server] Aborting
mysql         | 2020-11-09T17:17:35.224317Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.22)  MySQL Community Server - GPL.

**Comment by IoscaA** [comment]
I've the migration files in source/database/migrations

**Error:**
SQLSTATE[HY000] [2002] Connection refused (SQL: select * from information_schema.tables where table_schema = database and table_name = migrations and table_type = 'BASE TABLE')

**In the other terminal:**
mysql         | 2020-11-09 17:27:49+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mysql         | 2020-11-09 17:27:49+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.22-1debian10 started.
mysql         | 2020-11-09 17:27:49+00:00 [Note] [Entrypoint]: Initializing database files
mysql         | 2020-11-09T17:27:49.642046Z 0 [Warning] [MY-010915] [Server] 'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode. They will be merged with strict mode in a future release.
mysql         | 2020-11-09T17:27:49.642105Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.22) initializing of server in progress as process 42
mysql         | 2020-11-09T17:27:49.643103Z 0 [Warning] [MY-013242] [Server] --character-set-server: 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous.
mysql         | 2020-11-09T17:27:49.643715Z 0 [ERROR] [MY-010457] [Server] --initialize specified but the data directory has files in it. Aborting.
mysql         | 2020-11-09T17:27:49.643723Z 0 [ERROR] [MY-013236] [Server] The designated data directory /var/lib/mysql/ is unusable. You can remove all files that the server added to it.
mysql         | 2020-11-09T17:27:49.646801Z 0 [ERROR] [MY-010119] [Server] Aborting
mysql         | 2020-11-09T17:27:49.647503Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.22)  MySQL Community Server - GPL.

**Comment by supermavster** [comment]
Hello @IoscaA ,

1. Restore the mysql to the specific version of the project (https://github.com/supermavster/docker-laravel-8/)
2. Read again the **new** [README](../blob/main/README.md) and follow the steps.
3. My recommendation is delete all images and containers with the next commands: 

```sh
# Remove All
docker rm -vf $(docker ps -a -q)
docker rmi -f $(docker images -a -q)
docker system prune -f
docker volume prune -f
```


---
**NOTE:**

A little solution when you restore the project components is:

Clean the config with the new config in the files `.env` and `source/.env` with the next commands:

```
docker-compose run --rm artisan clear:data
docker-compose run --rm artisan cache:clear
docker-compose run --rm artisan view:clear
docker-compose run --rm artisan route:clear
docker-compose run --rm artisan clear-compiled
docker-compose run --rm artisan config:cache
```
And After that you need run your migrations
```
docker-compose run --rm artisan migrate --seed

```

**Comment by IoscaA** [comment]
I deleted everything about the old folder. I created the two .env files and after I run   **docker-compose run --rm composer create-project laravel/laravel .** This command produce this error:  **[InvalidArgumentException]                         
  Project directory "/var/www/html/." is not empty.**

This time I wanted to create a new laravel project instead of clone my personal project and follow your guide step by step. 

**Thanks for all :)**

**Comment by IoscaA** [comment]
After uninstalling and installing docker and docker-compose and adding my account to the docker group, I rerun everything and now I managed to create a new laravel project in the source folder. When I run _docker-compose run --rm artisan migrate --seed_ I obtain the same error :'c 

**SQLSTATE[HY000] [1130] Host 'docker-laravel-8_artisan_run_8bee1279a698.docker-laravel-8_l' is not allowed to connect to this MySQL server (SQL: select * from information_schema.tables where table_schema = db_name and table_name = migrations and table_type = 'BASE TABLE')**

I cleared the cache but it has no effect :(

**Comment by supermavster** [comment]
> I deleted everything about the old folder. I created the two .env files and after I run **docker-compose run --rm composer create-project laravel/laravel .** This command produce this error: **[InvalidArgumentException] Project directory "/var/www/html/." is not empty.**
> 
> This time I wanted to create a new laravel project instead of clone my personal project and follow your guide step by step.
> 
> **Thanks for all :)**

For this case in the README say: If exsit a errror delete this README.md, and try again.

Here is a example of this project: https://github.com/supermavster/laravel-docker-aws

**Comment by supermavster** [comment]
> After uninstalling and installing docker and docker-compose and adding my account to the docker group, I rerun everything and now I managed to create a new laravel project in the source folder. When I run _docker-compose run --rm artisan migrate --seed_ I obtain the same error :'c
> 
> **SQLSTATE[HY000] [1130] Host 'docker-laravel-8_artisan_run_8bee1279a698.docker-laravel-8_l' is not allowed to connect to this MySQL server (SQL: select * from information_schema.tables where table_schema = db_name and table_name = migrations and table_type = 'BASE TABLE')**
> 
> I cleared the cache but it has no effect :(

Check this source: https://github.com/supermavster/laravel-docker-aws and compare the files .env.example in the root and the source. After that you need use:


```
# Build
docker-compose up --build
# Clean Elements
docker-compose run --rm artisan clear:data
docker-compose run --rm artisan cache:clear
docker-compose run --rm artisan view:clear
docker-compose run --rm artisan route:clear
docker-compose run --rm artisan clear-compiled
docker-compose run --rm artisan config:cache
# Generate Key
docker-compose run --rm artisan key:generate
# Migrate
docker-compose run --rm artisan migrate --seed
```

**Comment by Hanschrs** [comment]
Hi @IoscaA I experienced the same problem here and realized that the mysql build process has shown error since we create the container. I did some workaround and it works:

1. remove the containers first
2. go to compose > mysql > my.cnf and edit the file
3. set innodb_force_recovery = 0
4. save
5. rebuild the container

I am still new to this, but by setting the innodb_force_recovery to 0 makes it work (tbh I do not know what will happen if I do so). But it works. Hope it helps.

**Comment by Hanschrs** [comment]
Hi @IoscaA I experienced the same problem here and realized that the mysql build process has shown error since we create the container. I did some workaround and it works:
1. remove the containers first
2. go to compose > mysql > my.cnf and edit the file
3. set innodb_force_recovery = 0
4. save
5. rebuild the container
I am still new to this, but by setting the innodb_force_recovery to 0 makes it work (tbh I do not know what will happen if I do so). But it works. Hope it helps.
